client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -166,73 +167,159 @@ public Response performRequest(String method, String endpoint, Map<String, Strin
+                               ResponseListener responseListener, Header... headers) {
         URI uri = buildUri(endpoint, params);
         HttpRequestBase request = createHttpRequest(method, uri, entity);
         setHeaders(request, headers);
-        //we apply a soft margin so that e.g. if a request took 59 seconds and timeout is set to 60 we don't do another attempt
-        long retryTimeoutMillis = Math.round(this.maxRetryTimeoutMillis / (float)100 * 98);
-        IOException lastSeenException = null;
+        FailureTrackingResponseListener failureTrackingResponseListener = new FailureTrackingResponseListener(responseListener);
         long startTime = System.nanoTime();
-        for (HttpHost host : nextHost()) {
-            if (lastSeenException != null) {
-                //in case we are retrying, check whether maxRetryTimeout has been reached, in which case an exception will be thrown
-                long timeElapsedMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
-                long timeout = retryTimeoutMillis - timeElapsedMillis;
-                if (timeout <= 0) {
-                    IOException retryTimeoutException = new IOException(
-                            "request retries exceeded max retry timeout [" + retryTimeoutMillis + "]");
-                    retryTimeoutException.addSuppressed(lastSeenException);
-                    throw retryTimeoutException;
+        performRequest(startTime, nextHost().iterator(), request, responseConsumer, failureTrackingResponseListener);
+    }
+
+    private void performRequest(final long startTime, final Iterator<HttpHost> hosts, final HttpRequestBase request,
+                                final HttpAsyncResponseConsumer<HttpResponse> responseConsumer,
+                                final FailureTrackingResponseListener listener) {
+        final HttpHost host = hosts.next();
+        //we stream the request body if the entity allows for it
+        HttpAsyncRequestProducer requestProducer = HttpAsyncMethods.create(host, request);
+        client.execute(requestProducer, responseConsumer, new FutureCallback<HttpResponse>() {
+            @Override
+            public void completed(HttpResponse httpResponse) {
+                try {
+                    RequestLogger.logResponse(logger, request, host, httpResponse);
+                    int statusCode = httpResponse.getStatusLine().getStatusCode();
+                    Response response = new Response(request.getRequestLine(), host, httpResponse);
+                    if (isSuccessfulResponse(request.getMethod(), statusCode)) {
+                        onResponse(host);
+                        listener.onSuccess(response);
+                    } else {
+                        ResponseException responseException = new ResponseException(response);
+                        if (mustRetry(statusCode)) {
```

Comment:
I usually think of retries in terms of "am I allowed to retry this exception if I want to" VS "this one means something is totally wrong and I shouldn't". I think that is how the http spec is worded too. Anyway, I think of that method in the same way. It is answering the question "would it be OK if I retried this request". Then other code makes a decision on whether or not it should retry based on things like if it has another node or not.